agent/consul/gateways/gateway_meta.go

```diff
@@ -0,0 +1,143 @@
+package gateways
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/consul/state"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// GatewayMeta embeds both a BoundAPIGateway and its corresponding APIGateway.
+type GatewayMeta struct {
+	// Bound is the bound config entry for a given gateway.
+	Bound *structs.BoundAPIGatewayConfigEntry
+	// Gateway is the APIGateway config entry for the gateway.
+	Gateway *structs.APIGatewayConfigEntry
+}
+
+// GetGatewayMeta queries the state store for an API Gateway and a Bound API
+// Gateway matching the given name and enterprise meta and returns a GatewayMeta
+// struct containing both.
+func GetGatewayMeta(store *state.Store, name string, entMeta *acl.EnterpriseMeta) (*GatewayMeta, error) {
+	_, bound, err := store.ConfigEntry(nil, structs.BoundAPIGateway, name, entMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	_, gateway, err := store.ConfigEntry(nil, structs.APIGateway, name, entMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &GatewayMeta{
+		Bound:   bound.(*structs.BoundAPIGatewayConfigEntry),
+		Gateway: gateway.(*structs.APIGatewayConfigEntry),
+	}, nil
+}
+
+// UpdateGatewayMeta takes a parent resource reference and a BoundRoute and
+// modifies the listeners on the BoundAPIGateway config entry in GatewayMeta
+// to reflect the binding of the route to the gateway.
+//
+// If the reference is not valid or the route's protocol does not match the
+// targeted listener's protocol, a mapping of parent references to associated
+// errors is returned.
+func (g *GatewayMeta) UpdateRouteBinding(refs []structs.ResourceReference, route structs.BoundRoute) (bool, map[structs.ResourceReference]error) {
```

Comment:
I didn't add tests for this directly because they would essentially repeat the tests for the `BindRoutesToGateways` function... That function is a kind of wrapper around this that filters out irrelevant binding cases before going through this process.